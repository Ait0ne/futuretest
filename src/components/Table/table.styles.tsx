import styled from 'styled-components';


export const TableContainer = styled.div`
display:flex;
flex-direction: column;
justify-content: center;
align-items: center;
width: 100%;
`

export const StyledTable = styled.table`
width:100%;
display: table;
font-family: 'Roboto', sans-serif;
border-spacing: 0;
border-collapse: collapse;
>thead>tr {
    border-bottom: 1px solid rgba(224, 224, 224, 1);
    >th {
    display: table-cell;
    padding: 12px;
}
}
`

interface TableHeadProps {
    sortDirection?: 'asc'|'desc'
}

export const TableHead = styled.th<TableHeadProps>`
cursor:pointer;
text-align: left;
text-transform: capitalize;
>svg {
    color: ${props=>props.sortDirection? 'rgb(17,17,17)' : 'rgba(17,17,17, 0.3)'};
    transition-duration: 0.3;
    transform: ${props => props.sortDirection==='asc'? 'rotate(180deg)':'none'}
}
`

interface TableRowProps {
    active?: boolean
}

export const TableRow = styled.tr<TableRowProps>`
cursor: pointer;
font-size: 0.875rem;
border-bottom: 1px solid rgba(224, 224, 224, 1);
background-color: ${props => props.active? 'rgba(0,0,0,0.1)': 'white'};
>td {
    display: table-cell;
    padding: 12px;
}
`

