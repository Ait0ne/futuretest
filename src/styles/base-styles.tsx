import styled from 'styled-components';

interface ButtonProps {
    active?: boolean
}

export const Button = styled.button<ButtonProps>`
cursor: pointer;
min-width: 64px;
padding: 6px 16px;
color: rgba(0, 0, 0, 0.87);
text-transform: uppercase;
font-weight: 500;
border-radius:4px;
border: none;
margin: 5px;
background-color: #e0e0e0;
box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14), 0px 1px 5px 0px rgba(0,0,0,0.12);
&:disabled {
    color: rgba(0, 0, 0, 0.26)
}
`

export const Input = styled.input`
padding: 12px 25px 12px 14px;
border-radius: 4px;
width:25ch;
border: 1px solid rgba(0,0,0,0.3);
font-size: 1rem;
margin: 5px 10px;
&:focus {
    outline:none;
}

`