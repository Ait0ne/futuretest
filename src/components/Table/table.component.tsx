import React from 'react';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faCaretDown} from '@fortawesome/free-solid-svg-icons';

import {TableContainer, StyledTable, TableHead, TableRow} from './table.styles';
import {ISort} from '../Main/main.component';




export interface TableProps {
    list: {[s:string]:any}[],
    sort: ISort|undefined,
    changeSortType: (newSort:ISort)=> void,
    columns: string[],
    changeSelectedRow: (rowData: any) => void,
    selectedRow: Object|undefined
}

const Table: React.FC<TableProps> = ({list, sort, changeSortType, columns, selectedRow, changeSelectedRow}) => {
    
    const sortTable = (column:string) => {
        let newSort:ISort;
        if (sort?.column===column) {
            newSort = {
                column,
                direction: sort?.direction==='asc'?'desc': 'asc'
            }
        } else {
            newSort = {
                column,
                direction: 'desc'
            }
        }
        changeSortType(newSort)
    }
    

    

    return(
        <TableContainer>
            <StyledTable>
                <thead>
                    <tr>
                        {
                            columns.map((column, index)=> {
                                return (
                                <TableHead
                                key={index}
                                onClick={() => sortTable(column)}
                                sortDirection={sort?.column===column? sort.direction: undefined}
                                >
                                    {column}<FontAwesomeIcon icon={faCaretDown}/>
                                </TableHead>
                                )
                            })
                        }
                    </tr>
                </thead>
                <tbody>
                    {
                        list.map((row, index)=>{
                            return (
                            <TableRow 
                            key={index}
                            active={JSON.stringify(row)===JSON.stringify(selectedRow)}
                            onClick={() => changeSelectedRow(row)}
                            >
                                {
                                    columns.map((column, index)=> {
                                        return (
                                            <td key={index}>{row[column]}</td>
                                        )
                                    })
                                }
                            </TableRow>
                            )
                        })
                    }
                </tbody>
            </StyledTable>
        </TableContainer>
    )
}

export default Table;
