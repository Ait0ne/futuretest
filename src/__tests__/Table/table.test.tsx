import React from 'react';
import {shallow} from 'enzyme';
import {TableColumns} from '../../components/Main/main.component';
import Table, {TableProps} from '../../components/Table/table.component';
import {testData} from '../../testData';
import {StyledTable, TableHead, TableRow} from '../../components/Table/table.styles';

const testProps:TableProps = {
    changeSortType: jest.fn(),
    columns: Object.keys(TableColumns),
    list: testData.slice(0,3),
    sort: undefined,
    changeSelectedRow: jest.fn(),
    selectedRow: undefined
}

describe('table', () => {
    const table = shallow(<Table {...testProps}/>)
    
    it('renders correctly', () => {
        expect(table).toMatchSnapshot()
    })

    it('renders one table', () => {
        expect(table.find(StyledTable).length).toEqual(1)
    })

    it('has correct number of columns', () => {
        expect(table.find(TableHead).length).toEqual(testProps.columns.length)
    })    

    it('renders correct number of rows', () => {
        expect(table.find(TableRow).length).toEqual(testProps.list.length)
    })

    describe('click on columns header', () => {
        const header = table.find(TableHead).at(0)
        header.simulate('click')
         
        it('calls sort function with correct values', ()=> {
            expect(testProps.changeSortType).toHaveBeenCalledWith({column: testProps.columns[0], direction: 'desc'})
        })
    })

    describe('click on a row', () => {
        const row = table.find(TableRow).at(0)
        row.simulate('click')

        it('calls changeSelectedRow function with correct values', () => {
            expect(testProps.changeSelectedRow).toHaveBeenCalledWith(testData[0])
        })
    })

})

describe('table with different props', () => {
    const table = shallow(<Table {...{...testProps, sort: {column: 'id', direction: 'desc'}}}/>)
    
    describe('column header', () => {
        const header = table.find(TableHead).at(0)
        header.simulate('click')
         
        it('has correct sortDirectionProp', () => {
            expect(header.prop('sortDirection')).toEqual('desc')
        })


        it('calls sort function with correct values', ()=> {
            expect(testProps.changeSortType).toHaveBeenCalledWith({column: testProps.columns[0], direction: 'asc'})
        })
    })
})



