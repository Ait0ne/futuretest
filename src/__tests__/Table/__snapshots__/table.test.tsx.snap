// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table renders correctly 1`] = `
<styled.div>
  <styled.table>
    <thead>
      <tr>
        <styled.th
          key="0"
          onClick={[Function]}
        >
          id
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  320,
                  512,
                  Array [],
                  "f0d7",
                  "M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z",
                ],
                "iconName": "caret-down",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.th>
        <styled.th
          key="1"
          onClick={[Function]}
        >
          firstName
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  320,
                  512,
                  Array [],
                  "f0d7",
                  "M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z",
                ],
                "iconName": "caret-down",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.th>
        <styled.th
          key="2"
          onClick={[Function]}
        >
          lastName
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  320,
                  512,
                  Array [],
                  "f0d7",
                  "M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z",
                ],
                "iconName": "caret-down",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.th>
        <styled.th
          key="3"
          onClick={[Function]}
        >
          email
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  320,
                  512,
                  Array [],
                  "f0d7",
                  "M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z",
                ],
                "iconName": "caret-down",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.th>
        <styled.th
          key="4"
          onClick={[Function]}
        >
          phone
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  320,
                  512,
                  Array [],
                  "f0d7",
                  "M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z",
                ],
                "iconName": "caret-down",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.th>
      </tr>
    </thead>
    <tbody>
      <styled.tr
        active={false}
        key="0"
        onClick={[Function]}
      >
        <td
          key="0"
        >
          599
        </td>
        <td
          key="1"
        >
          Ning
        </td>
        <td
          key="2"
        >
          Sherman
        </td>
        <td
          key="3"
        >
          TGalvez@vitae.org
        </td>
        <td
          key="4"
        >
          (906)905-9693
        </td>
      </styled.tr>
      <styled.tr
        active={false}
        key="1"
        onClick={[Function]}
      >
        <td
          key="0"
        >
          594
        </td>
        <td
          key="1"
        >
          Sanjeev
        </td>
        <td
          key="2"
        >
          Settle
        </td>
        <td
          key="3"
        >
          SBarrientos@sollicitudin.ly
        </td>
        <td
          key="4"
        >
          (716)271-5941
        </td>
      </styled.tr>
      <styled.tr
        active={false}
        key="2"
        onClick={[Function]}
      >
        <td
          key="0"
        >
          364
        </td>
        <td
          key="1"
        >
          Lynda
        </td>
        <td
          key="2"
        >
          Lompa
        </td>
        <td
          key="3"
        >
          SSchamber@lacus.com
        </td>
        <td
          key="4"
        >
          (948)483-7779
        </td>
      </styled.tr>
    </tbody>
  </styled.table>
</styled.div>
`;
